xtask/src/semver_check.rs

```diff
@@ -0,0 +1,257 @@
+use std::{
+    fs,
+    io::Write,
+    path::{Path, PathBuf},
+};
+
+use cargo_semver_checks::{Check, GlobalConfig, ReleaseType, Rustdoc};
+use esp_metadata::Chip;
+use rustdoc_types::ItemEnum;
+
+use crate::{Package, cargo::CargoArgsBuilder};
+
+/// Return the minimum required bump for the next release.
+/// Even if nothing changed this will be [ReleaseType::Patch]
+pub fn minimum_update(
+    workspace: &Path,
+    package: Package,
+    chip: Chip,
+) -> Result<ReleaseType, anyhow::Error> {
+    log::info!("Chip = {}", chip.to_string());
+
+    let package_name = package.to_string();
+    let package_path = crate::windows_safe_path(&workspace.join(&package_name));
+    let current_path = build_doc_json(package, &chip, &package_path)?;
+    remove_unstable_items(&current_path)?;
+
+    let file_name = if package.chip_features_matter() {
+        format!("{}", chip.to_string())
+    } else {
+        "api".to_string()
+    };
+
+    let baseline_path_gz =
+        PathBuf::from(&package_path).join(format!("api-baseline/{}.json.gz", file_name));
+    let baseline_path = temp_file::TempFile::new()?;
+    let buffer = Vec::new();
+    let mut decoder = flate2::write::GzDecoder::new(buffer);
+    decoder.write_all(&(fs::read(&baseline_path_gz)?))?;
+    fs::write(baseline_path.path(), decoder.finish()?)?;
+
+    let mut semver_check = Check::new(Rustdoc::from_path(current_path));
+    semver_check.set_baseline(Rustdoc::from_path(baseline_path.path()));
+    let mut cfg = GlobalConfig::new();
+    cfg.set_log_level(Some(log::Level::Info));
+    let result = semver_check.check_release(&mut cfg)?;
+    log::info!("Result {:?}", result);
+
+    let mut min_required_update = ReleaseType::Patch;
+    for (_, report) in result.crate_reports() {
+        if let Some(required_bump) = report.required_bump() {
+            min_required_update = match required_bump {
+                ReleaseType::Major
+                    if min_required_update == ReleaseType::Patch
+                        || min_required_update == ReleaseType::Minor =>
+                {
+                    ReleaseType::Major
+                }
+                ReleaseType::Minor if min_required_update == ReleaseType::Patch => {
+                    ReleaseType::Minor
+                }
+                _ => ReleaseType::Patch,
+            }
+        }
+    }
+
+    Ok(min_required_update)
+}
+
+pub(crate) fn build_doc_json(
+    package: Package,
+    chip: &Chip,
+    package_path: &PathBuf,
+) -> Result<PathBuf, anyhow::Error> {
+    let target_dir = std::env::var("CARGO_TARGET_DIR");
+
+    let target_path = if let Ok(target) = target_dir {
+        PathBuf::from(target)
+    } else {
+        PathBuf::from(package_path).join("target")
+    };
+    let current_path = target_path
+        .join(chip.target())
+        .join("doc")
+        .join(format!("{}.json", package.to_string().replace("-", "_")));
+
+    std::fs::remove_file(&current_path).ok();
+    let toolchain = "esp";
```

Comment:
Not a major thing, but this should be inlined into the `toolchain` call. There's no reason to place the value a good distance away from the call, especially when the variable name is the same as the function.